### PR TITLE
feat(frontend): Token group init

### PR DESCRIPTION
**Pull request by darioloetsc**
# Motivation

Extended token types to handle token group implementation.
Here the foundation is created to map the filtered tokens into groups or standalone tokens as is, to have a view render a group entry or the existing one based on matching. The types are prepared to implement the visual components.

# Changes

- Add types
- Add mapping function

# Tests

Add vitest for mapping function.

### CODE REVIEW

File: src/frontend/src/lib/types/token.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/types/token.ts:74]: `nativeNetwork`?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:195]: here it would be better to use the component that is a placeholder for logo, I think it is `Logo` or `IconRandom`. Or maybe it is not even necessary to add a fallback, since it is handled in the card (we usually don't add an icon fallback)

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:192]: why not using directly the network name that is defined instead of capitalize?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:194]: is this right? does the twinToken always have the same decimals?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:216]: may you please make it functional, maybe using reduce to define both variables `groupedTokenTwins` and `mappedTokensWithGroups`

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:221]: i would avoid casting: we should find a way to set the type of this variable and then build it accordingly

`const twinToken: IcCkToken | undefined = ....`

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:225]: `nonNullish(twinToken)`

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:229]: same comment for casting as above

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:237]: so, to reduce the loops, you could make the entire thing functional and apply the filter inside via logic

**Review by AntonioVentilii-DFINITY — COMMENTED** [review]
Just a first general comment: may you please add an explanation of what are the plans for the grouping, to give some context on what you intend of doing with the new types, just some overview

File: src/frontend/src/icp/utils/icrc.utils.ts
Comment (peterpeterparker) [reply src/frontend/src/icp/utils/icrc.utils.ts:155]: @AntonioVentilii-DFINITY can you create an eslint for enforing arrow function?

File: src/frontend/src/lib/types/token.ts
Comment (peterpeterparker) [reply src/frontend/src/lib/types/token.ts:70]: Generally speaking, why do we need a new UI type in this PR? Is that related to the grouping?

File: src/frontend/src/lib/types/token.ts
Comment (peterpeterparker) [reply src/frontend/src/lib/types/token.ts:74]: - `metadata`?
- we need the all nerwork information - not just the ID?

File: src/frontend/src/lib/types/token.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/types/token.ts:78]: @AntonioVentilii-DFINITY do we have a TS utils for one or the other already? If not, I think I have that somewhere (for another PR)

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/utils/token.utils.ts:188]: Is the `toUpperCase` meant for display purpose?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [reply src/frontend/src/lib/utils/token.utils.ts:221]: nitpick: we are fan of destructing everything so the code based mostly use this approach. not against a bit of consistency but for sure I'm aware this is an opiniated feedback so feel free to ignore.

File: src/frontend/src/lib/types/token.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/token.ts:78]: I remember that too, maybe I saw it in ic-js or gix-component... I will add it here and modify (I can do it in a separate  PR, so that it justifies why I am adding it)

File: src/frontend/src/icp/utils/icrc.utils.ts
Comment (darioloetsc) [reply src/frontend/src/icp/utils/icrc.utils.ts]: 4d933789

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: e75eaab0

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 0939339b

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: d6bbb2bc

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 7bcf93a7

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 7bcf93a7

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 7bcf93a7

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts:182]: Yes - prep to integrate the design, probably will rework it when implementing the view components

Comment (peterpeterparker) [reply src/frontend/src/lib/utils/token.utils.ts:182]: I see. Please remove this then. Capitalize the first letter should be solved with CSS:

Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: af243a0b

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: d6bbb2bc

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 7bcf93a7

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts:245]: i kept the list in full and then removed what is duplicated to have the order secured. Please propose code change if you see how it can be done better.

File: src/frontend/src/lib/types/token.ts
Comment (darioloetsc) [reply src/frontend/src/lib/types/token.ts:78]: Sure - buts lets progress here, have 3 follow up prs depending on it for now.

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:159]: I am not really sure I understand what is the use case here: why the input is `unkown` type? where is this one expected to be used? can we not limit the possible inputs?

File: src/frontend/src/lib/types/token.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/token.ts:70]: @darioloetsc any news on the above?

File: src/frontend/src/lib/types/token.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/token.ts:74]: @darioloetsc any news on the above?

File: src/frontend/src/lib/types/token.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/types/token.ts:78]: So, I will do it after in another PR and change it in this part? May you please ask Stefan to add it to the tasks? Thank you!!!

File: src/frontend/src/tests/lib/utils/token.utils.spec.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/tests/lib/utils/token.utils.spec.ts:24]: @darioloetsc any news here?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:186]: here it would be better to use the component that is a placeholder for logo, I think it is `Logo` or `IconRandom`. Or maybe it is not even necessary to add a fallback, since it is handled in the card (we usually don't add an icon fallback)

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:185]: is this right? does the `twinToken` always have the same decimals?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/utils/token.utils.ts:243]: have you tried to inject the filter in `reduce`? We have some examples in the code with functions `pinTokensWithBalanceAtTop` and `sumMainnetTokensUsdBalancesPerNetwork`

File: src/frontend/src/lib/types/token.ts
Comment (darioloetsc) [reply src/frontend/src/lib/types/token.ts:70]: oh yes, thats used for a follow up commit, sry sliced it wrongly

**Comment by AntonioVentilii-DFINITY** [comment]
@darioloetsc please check my comment on the description too https://github.com/dfinity/oisy-wallet/pull/2969#pullrequestreview-2374573041

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 459473a4

File: src/frontend/src/lib/types/token.ts
Comment (darioloetsc) [reply src/frontend/src/lib/types/token.ts]: 459473a4

yes, the network information is needed in the view component later on

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 5664d75a

File: src/frontend/src/tests/lib/utils/token.utils.spec.ts
Comment (darioloetsc) [reply src/frontend/src/tests/lib/utils/token.utils.spec.ts:24]: Good call - somehow missed that, however twinTokenSymbol: 'ckBTC' is missing on that declaration, which should be there if compared to ETHEREUM_TOKEN. Any idea why?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: removed anyway

File: src/frontend/src/tests/lib/utils/token.utils.spec.ts
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/tests/lib/utils/token.utils.spec.ts:24]: not really sure to be honest, but I think it has to do with that big tasks that is conversion `BTC <--> ckBTC`, so, if I were you, i would ignore and inject only in the test manually, for now

maybe add a TODO "remove this when ckBTC will be added, blah blah blah..."

File: src/frontend/src/lib/types/token.ts
Comment (peterpeterparker) [reply src/frontend/src/lib/types/token.ts:70]: > thats used for a follow up commit

I'm confuse, so we do not need it here and it's not related to grouping?

File: src/frontend/src/icp/utils/icrc.utils.ts
Comment (peterpeterparker) [inline src/frontend/src/icp/utils/icrc.utils.ts:155]: `ICP_NETWORK_ID`?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/utils/token.utils.ts:161]: `!== null` instead of `isNullish` because TS is not able to  interpret it or because `undefined` is valid?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [reply src/frontend/src/lib/utils/token.utils.ts:185]: I think so but, that's a good point, maybe not that future proof knowing how many times we already had to refactor the core. 

Maybe when we group tokens, if their decimals is different, it should not be possible to group them? Or better idea?

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/utils/token.utils.ts:213]: We generally prefer a functional approach. Not necesseraly for this PR but, maybe add a TODO to review this.

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/utils/token.utils.ts:217]: Curiosity question, declaring the type here `TokenUi | undefined` is required because TS cannot interfer it?

File: src/frontend/src/lib/types/token.ts
Comment (darioloetsc) [reply src/frontend/src/lib/types/token.ts:78]: Reported

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 01f16459

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 5664d75a

File: src/frontend/src/icp/utils/icrc.utils.ts
Comment (darioloetsc) [reply src/frontend/src/icp/utils/icrc.utils.ts]: 3e404bc8

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 6af63074

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: 83a1d35f

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: d8cdc227 here a refactored version which is more functional style

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts:185]: To play it safe should make sense here. Here a version which doesn't group on mismatch: b233fe93

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: b233fe93 here a refactored version

**Comment by darioloetsc** [comment]
> Just a first general comment: may you please add an explanation of what are the plans for the grouping, to give some context on what you intend of doing with the new types, just some overview

Here the foundation is created to map the filtered tokens into groups or standalone tokens as is, to have a view render a group entry or the existing one based on matching. The types are prepared to implement the visual components.

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:161]:
```suggestion
	nonNullish(tokenUiOrGroupUi) &&
```

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:206]: nitpick: we usually initialize the type of the variables of `reduce` directly here, for example

`reduce<{token: ... ; groupedTokenSymbols: ...}>`

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:221]: `groupedTokenSymbols: [...acc.groupedTokenSymbols, token.symbol, twinToken.symbol]`

WDYT?

File: src/frontend/src/tests/lib/utils/token.utils.spec.ts
Comment (darioloetsc) [reply src/frontend/src/tests/lib/utils/token.utils.spec.ts]: I extended the declared type, not having that value could break grouping. It shouldn't harm and now has the same declaration as eth and usdc. 27aab1cc

File: src/frontend/src/lib/utils/token.utils.ts
Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: Playing safe b233fe93

File: src/frontend/src/lib/utils/token.utils.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/utils/token.utils.ts:204]: right now this function suppose that the ckTokens comes always after their twin counterparty in the order of `tokens` input (that is generally the case, I suppose?): if a ckToken appears first in the list, what happens?

Furthermore, it considers only groups of two by two: for example, if, for whatever reason we have ETH, ckETH and ETH-BNB, that refers all to ETH, they will not be grouped. That is ok for now since it is not needed.

However, in the interest of time and just to track it, may you please add a TODO here to mention both issues above?

**Review by AntonioVentilii-DFINITY — APPROVED** [review]
Please have a look at my last comments before merging, but already approving

However, please wait for @peterpeterparker to approve too, since he reviewed it too

File: src/frontend/src/lib/types/token.ts
Comment (Stefan-Berger-DFINITY) [reply src/frontend/src/lib/types/token.ts:76]: https://dfinity.atlassian.net/browse/OISY-141

File: src/frontend/src/lib/utils/token.utils.ts
Comment (peterpeterparker) [reply src/frontend/src/lib/utils/token.utils.ts:204]: We have changed the order of the tokens countless times, so for future-proofing, it might be better to implement logic that makes no assumptions about the ordering.

Regarding the second point, we discussed this week that SepoliaETH might eventually be grouped with ETH. Therefore, I believe it would be future-proof not to assume that there are only two tokens per group.

Comment (peterpeterparker) [reply src/frontend/src/lib/utils/token.utils.ts:204]: Given that there were already too many comments in the PR, I suggest we add a TODO and address those questions in follow-up PRs. I mean, not something we schedule, but right after we merge this one.

Comment (darioloetsc) [reply src/frontend/src/lib/utils/token.utils.ts]: Yes, that bug was introduced. I reverted it to the easy implementation: ed145f4d. I will ping Stefan to do a tech debt ticket to reimplement the function in an ideologically correct way.

Comment (peterpeterparker) [reply src/frontend/src/lib/utils/token.utils.ts]: No need to add it to debt, it's a bug we want to fix before using it. In another PR but, we gonna fix that before any usage.

Comment (Stefan-Berger-DFINITY) [reply src/frontend/src/lib/utils/token.utils.ts]: https://dfinity.atlassian.net/browse/OISY-142

Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/utils/token.utils.ts]: If i may give a suggestion on implementation: I would avoid creating a list of mixed tokens and grouped tokens.
I think it is to return a list of just groups, and in some of those groups there is only a single token.

Then you handle differently single-element groups from multiple-elements groups, but you do it elsewhere.

Doing so will simplify the grouping function quite a bit, I think, and it would use only one `reduce` with no additional loops.

**Review by peterpeterparker — APPROVED** [review]
I'm approving this PR to move forward, but I'm considering the following leftovers discussed in this [thread](https://github.com/dfinity/oisy-wallet/pull/2969#discussion_r1805966235) as must-haves before effectively implementing the "grouping" feature:

- We should fix the assumption that ckTokens always come after their twin counterpart, as this could potentially lead to missing or duplicated tokens in the application.

- We want to remove the limitation of two tokens per group, given that we already discussed the possibility of having more than two tokens per group, such as with SepoliaETH. Based on our experience, adding technical debt to the core features — the tokens — of the application is a killer of time and generates issues - even in production given the lack of tests - when new requirements need to be implemented.

- To implement these improvements and align with the architecture and team's programming approach, I warmly suggest following a functional programming approach. This can potentially ease collaboration and maintain the consistency of the product's codebase.

On a final note, while writing these lines, I noticed that @AntonioVentilii-DFINITY added a final [note](https://github.com/dfinity/oisy-wallet/pull/2969#discussion_r1806071461) that might offer a solution to potentially resolve all the issues. Please reach out to him for further discussion.

Thanks!

**Pull request merged**
